File: solutions/c/resistor-color-trio/1/resistor_color_trio.c
#include "resistor_color_trio.h"

#include <stdio.h>
#include <inttypes.h>
#include <math.h>

resistor_value_t color_code(resistor_band_t bands[]) {
    uint32_t value = (10 * bands[0] + bands[1]) * pow(10, bands[2]);
    unit_t unit = OHMS;

    if (value >= 1000) {
        value /= 1000;
        unit = KILOOHMS;
    }

    return (resistor_value_t) { value, unit };
}

File: solutions/go/resistor-color-trio/1/resistor_color_trio.go
package resistorcolortrio

import (
	"fmt"
	"math"
)

const (
	giga = 1000000000
	mega = 1000000
	kilo = 1000
)

var resistances = map[string]int{
	"black":  0,
	"brown":  1,
	"red":    2,
	"orange": 3,
	"yellow": 4,
	"green":  5,
	"blue":   6,
	"violet": 7,
	"grey":   8,
	"white":  9,
}

// Label returns the resistance value for a given resistor label.
func Label(colors []string) string {
	resistance := (resistances[colors[0]]*10 + resistances[colors[1]]) * int(math.Pow(10, float64(resistances[colors[2]])))
	metric := "ohms"

	if resistance/giga > 0 {
		resistance /= giga
		metric = "gigaohms"
	} else if resistance/mega > 0 {
		resistance /= mega
		metric = "megaohms"
	} else if resistance/kilo > 0 {
		resistance /= kilo
		metric = "kiloohms"
	}

	return fmt.Sprintf("%d %s", resistance, metric)
}
